src/mawpy/steps/address_oscillation.py

```diff
@@ -0,0 +1,254 @@
+import logging
+
+import numpy as np
+import pandas as pd
+
+from mawpy.constants import USER_ID, UNIX_START_T, STAY_DUR, STAY_LAT, ORIG_LAT, ORIG_LONG, STAY_LONG, AO_COLUMNS
+from mawpy.utilities.preprocessing import get_preprocessed_dataframe, get_list_of_chunks_by_column, execute_parallel
+
+logger = logging.getLogger(__name__)
+
+
+def _transform_trace(row: pd.Series, df_columns: list) -> pd.Series:
+    """
+    Transforms a row of the dataframe by determining the duration, latitude, longitude,
+    and whether it's a stay point or not.
+
+    Args:
+        row (pd.Series): A row from the dataframe.
+
+    Returns:
+        pd.Series: Transformed row with duration, latitude, longitude, and stay/not stay indicator.
+    """
+    # Determine duration, defaulting to 1 if not present or stay_dur
+    duration = 1 if STAY_DUR not in df_columns or row[STAY_DUR] == -1 else row[STAY_DUR]
+
+    # Determine latitude, defaulting to original if not present or stay_lat
+    lat = row[ORIG_LAT] if STAY_LAT not in df_columns or row[STAY_LAT] == -1 else row[STAY_LAT]
+
+    # Determine longitude, defaulting to original if not present or stay_long
+    long = row[ORIG_LONG] if STAY_LONG not in df_columns or row[STAY_LONG] == -1 else row[STAY_LONG]
+
+    # Determine if it's a stay point, defaulting to 0 if not present or 1
+    stay_not = 0 if STAY_LAT not in df_columns or row[STAY_LAT] == -1 else 1
+
+    return pd.Series([duration, lat, long, stay_not])
+
+
+def _get_stays_by_lat_long_dur(df_by_user: pd.DataFrame) -> np.ndarray:
+    """
+    Groups together consecutive traces that have the same value for latitude, longitude, and duration.
+
+    Args:
+        df_by_user (pd.DataFrame): Dataframe filtered by user.
+
+    Returns:
+        np.ndarray: Array of stay group indices.
+    """
+    lat = df_by_user['lat'].to_numpy()
+    long = df_by_user['long'].to_numpy()
+    duration = df_by_user['duration'].to_numpy()
+
+    n = len(lat)
+    stays = np.zeros(n, dtype=int)
+    current_stay = 0
+
+    for i in range(1, n):
+        # Increment stay group if lat, long, or duration changes
+        if lat[i] != lat[i - 1] or long[i] != long[i - 1] or duration[i] != duration[i - 1]:
```

Comment:
Are these floats? If so, do these comparisons need any tolerance? Ie, do we need to use something like `numpy.isclose()` or is exact float equality desired?

Reply:
These are coming from columns from the input data frame (that is up to 6 decimal points), so I believe the exact check should be fine.